inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_buffer_fusing.cpp

```diff
@@ -38,217 +38,242 @@
+    using base = pattern_match_optimization_typed<concat_in_place_optimization, concatenation>;
+    using base::base;
+
+    // Runs concat in-place optimization and adds already optimized concatenations that need re-optimization to `needs_reoptimization`.
+    void optimize_cascade(concatenation_node& node, std::list<concatenation_node*>& need_reoptimization);
+    bool match(concatenation_node& node);
+    bool optimize(concatenation_node& node) {
+        std::list<concatenation_node*> need_reopt;
+        optimize_cascade(node, need_reopt);
+        while (!need_reopt.empty()) {
+            auto& prop = *need_reopt.front();
+            need_reopt.pop_front();
+            if (match(prop))
+                optimize_cascade(prop, need_reopt);
+            else
+                // TODO: Revert extra padding when cascade adjustment failed.
+                prop.can_be_optimized(false);
         }
-        return true;
-    };
+        return false;  // node not invalidated
+    }
+};
 
-    // [1] First try to optimize all concats
-    auto node_itr = p.get_processing_order().begin();
-    while (node_itr != p.get_processing_order().end()) {
-        auto& node = (*node_itr++);
-        if (!can_optimize(node))
-            continue;
-        program_helpers::do_for_types<concatenation>(*node, [&p, is_debug](concatenation_node& node) {
-            // For in place concatenation input layouts and data types must match
-            auto output_format = node.get_output_layout().format;
-            auto output_datatype = node.get_output_layout().data_type;
-            // we need to avoid mixing padded and unpadded buffer
-            bool all_dependencies_padded = true;
-            bool all_dependencies_unpadded = true;
-            for (auto& input : node.get_dependencies()) {
-                if (input->type() == reshape::type_id())
-                    // reshapes should be optimized out
-                    return;
+bool concat_noop_optimization::match(concatenation_node& node) {
+    return node.get_dependencies().size() == 1 &&
+        !node.has_fused_primitives() &&
+        node.get_fused_activations_funcs().empty();
+}
 
-                layout l = input->get_output_layout();
-                if (static_cast<bool>(l.data_padding))
-                    all_dependencies_unpadded = false;
-                else
-                    all_dependencies_padded = false;
+bool concat_noop_optimization::optimize(concatenation_node& node) {
+    auto& dep = node.get_dependency(0);
+    dep.merge_output_padding(node.get_output_layout().data_padding);
+    prog.extract_and_remove(node);
+    // Node has been removed, so no further optimizations.
+    return true;
+}
 
-                if (output_format != l.format || output_datatype != l.data_type)
-                    return;
+bool concat_in_place_optimization::match(concatenation_node& node) {
+    if (node.has_fused_primitives() || !node.get_fused_activations_funcs().empty())
```

Comment:
As I can see, there's no `node.is_output()` check anymore. Is it safe to do this transform for concat which is output?

Reply:
Moved to actual optimizations, previously it was in base class.